string/strerror.c

```diff
@@ -65,25 +66,59 @@ static inline const char *strerror_(const errinfo_t *tab, size_t tabsz, const ch
 	}
 
 	e = bsearch((void *)(long)errnum, tab, tabsz / sizeof(*tab), sizeof(*tab), err_cmp);
-
 	if (e == NULL) {
-		(void)sprintf(err_unknownMsg, "Unknown error %d", errnum);
-		errno = EINVAL;
-		return err_unknownMsg;
-	}
-	else {
-		return names + e->offset;
+		(void)snprintf(buff, bufflen, "Unknown error %d", errnum);
+		*err = EINVAL;
+		return buff;
 	}
+
+	return names + e->offset;
 }
 
 
 char *strerror(int errnum)
 {
-	return (char *)strerror_(errtab, sizeof(errtab), errnames, errnum);
+	const char *str;
+	int err = 0;
+
+	str = strerror_(errtab, sizeof(errtab), errnames, errnum, err_unknownMsg, sizeof(err_unknownMsg), &err);
+	if (err != 0) {
+		errno = err;
+	}
+
+	return (char *)str;
+}
+
+
+int strerror_r(int errnum, char *buf, size_t buflen)
+{
+	const char *str;
+	char unknownMsg[32];
+	int ret = 0;
+	size_t len;
+
+	str = strerror_(errtab, sizeof(errtab), errnames, errnum, unknownMsg, sizeof(unknownMsg), &ret);
+	len = strlen(str);
+	if (buflen > len) {
+		memcpy(buf, str, len + 1);
+	}
+	else {
+		ret = ERANGE;
+	}
+
+	return ret;
 }
 
 
 const char *gai_strerror(int errnum)
 {
-	return strerror_(gaitab, sizeof(gaitab), gainames, errnum);
+	const char *str;
+	int err;
```

Comment:
initialization of the err variable is missing
```suggestion
	int err = 0;
```